2011-02-12  Alexandre Oliva  <aoliva@redhat.com>

	PR debug/47620
	* haifa-sched.c (fix_tick_ready): Don't queue debug insns.

--- gcc/haifa-sched.c.orig	2011-02-12 10:26:29.000000000 -0200
+++ gcc/haifa-sched.c	2011-02-12 10:39:17.584244783 -0200
@@ -3929,6 +3929,15 @@ fix_tick_ready (rtx next)
 	  if (!full_p)
 	    break;
         }
+
+      /* We can't queue debug insns, but we would try to during
+	 init_ready_list when a debug insn is the very first.  Arrange
+	 for it to be added to the ready list instead.  */
+      if (!tick && clock_var == -1 && DEBUG_INSN_P (next))
+	{
+	  gcc_checking_assert (q_size == 0);
+	  tick = -1;
+	}
     }
   else
     tick = -1;
